File: multMatrix.py
import time
import numpy as np

def read_matrix(filename, n):
    matrix = []
    with open(filename, 'r') as file:
        for _ in range(n):
            row = list(map(float, file.readline().split()))
            matrix.append(row)
    return matrix

def write_time_to_file(filename, time_taken, n):
    with open(filename, 'a') as f:
        f.write(f"\nTime taken for {n}x{n} Matrix in Python: {time_taken:.4f} seconds\n")

def main():
    n = 500
    matrix1 = read_matrix("./sample matrix/matrix_500A.txt", n)
    matrix2 = read_matrix("./sample matrix/matrix_500B.txt", n)

    # Convert to numpy arrays for fast multiplication
    mat1_np = np.array(matrix1)
    mat2_np = np.array(matrix2)

    # Time the multiplication
    start_time = time.time()
    result = np.matmul(mat1_np, mat2_np)
    end_time = time.time()

    time_taken = end_time - start_time
    write_time_to_file("timeTakenPy.txt", time_taken, n)

if __name__ == "__main__":
    main()

File: multMatrix.c
#include <stdio.h>
#include <time.h>
#include "./headers/operations.h"
#include "./operations/multMatrix.c"
#include "./util/allocateMatrix.c"

int main()
{

    int n = 500;
    float **toM1 = allocateMatrix(n, n);
    float **toM2 = allocateMatrix(n, n);
    ;
    float **rm = allocateMatrix(n, n);
    ;
    FILE *file1 = fopen("./sample matrix/matrix_500A.txt", "r");
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            fscanf(file1, "%d", &toM1[i][j]);
    FILE *file2 = fopen("./sample matrix/matrix_500B.txt", "r");
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            fscanf(file2, "%d", &toM2[i][j]);
    clock_t start = clock();
    multMatrix(n, n, n, toM1, toM2, rm);
    clock_t end = clock();
    double time_taken = ((double)(end - start));
    FILE *toWrite = fopen("timeTakenC.txt", "a");
    fprintf(toWrite, "\nTime taken for 500x500 Matrix in C: %.4f seconds\n", time_taken);
    fclose(file1);
    fclose(file2);
}